exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,246 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+
+	for i := range machines {
+		machine := &machines[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		// We want to delete since an empty infraRef is invalid as well.
+		if (infraRef == corev1.ObjectReference{}) {
+			log.V(2).Info("Machine has an empty infraRef, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(2).Info("Machine is not orphaned, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine, infraMachines []unstructured.Unstructured) error {
+	infraRefNames := sets.Set[string]{}
+	for _, machine := range machines {
+		infraRef := machine.Spec.InfrastructureRef
+		infraRefNames.Insert(infraRef.Name)
+	}
+
+	for i := range infraMachines {
+		infraMachine := &infraMachines[i]
+		if infraRefNames.Has(infraMachine.GetName()) {
+			continue
+		}
+
+		machine := getNewMachine(mp, infraMachine)
+		if err := r.Client.Create(ctx, machine); err != nil {
+			return errors.Wrapf(err, "failed to create new Machine for infraMachine %q in namespace %q", infraMachine.GetName(), infraMachine.GetNamespace())
+		}
+	}
+
+	return nil
+}
+
+// setInfraMachineOwnerRefs creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) setInfraMachineOwnerRefs(ctx context.Context, mp *expv1.MachinePool, infraMachines []unstructured.Unstructured) error {
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		clusterv1.MachinePoolNameLabel: mp.Name,
+		clusterv1.ClusterNameLabel:     mp.Spec.ClusterName,
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
```

Comment:
I think it's possible but will be a bit more complicated since this function needs to do two things -- (1) set the owner refs for machines that just got created and (2) set the owner refs on existing machines in case it failed in step 1. If we want to reuse the list, we would need to also change the logic of `createMachinesIfNotExists` to attempt to set the owner ref too. So the trade off there is that the logic would exist in two places instead of one helper. Do you think we should proceed?

Reply:
1 is not necessary if you create machines with owner ref already set

Reply:
Can we just add machines being created to machineList?

Reply:
> set the owner refs for machines that just got created

aren't we already setting the owner refs when creating them?

Reply:
My mistake -- I meant to say that this code needs to set the ownerRef of the InfraMachines to its corresponding Machine. When the InfraMachine is created, it's owned by the InfraMP since the Machine doesn't exist yet.

Reply:
@fabriziopandini I suppose we can, if we want to maintain our own running copy of the Machines in the loop, we'd just have to be careful to make sure it doesn't get out of sync. I figured it would be less error prone to requery.